pkg/reconciler/workload/apply.go

```diff
@@ -40,52 +36,41 @@ var (
 	errPatchKubeApp          = "cannot patch existing KubernetesApplication"
 )
 
-// KubeAppApply applies changes to the supplied KubernetesApplication. The
-// KubernetesApplication will be created if it does not exist, or patched if it
-// does. Resource templates are merged instead of replaced if they have the same
-// name and GroupVersionKind.
-func KubeAppApply(ctx context.Context, c client.Client, o runtime.Object, ao ...resource.ApplyOption) error {
-	opts := &resource.ApplyOptions{}
-	for _, fn := range ao {
-		fn(opts)
-	}
-
-	m, ok := o.(*workloadv1alpha1.KubernetesApplication)
-	if !ok {
-		return errors.New(errNotKubeApp)
-	}
-
-	desired := &workloadv1alpha1.KubernetesApplication{}
-	m.DeepCopyInto(desired)
-
-	err := c.Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, o)
-	if kerrors.IsNotFound(err) {
-		return errors.Wrap(c.Create(ctx, o), errCreateKubeApp)
-	}
-	if err != nil {
-		return errors.Wrap(err, errGetKubeApp)
-	}
-
-	if opts.ControllersMustMatch && !meta.HaveSameController(m, desired) {
-		return errors.New(errDiffController)
-	}
-
-	for i, temp := range m.Spec.ResourceTemplates {
-		if desired.Spec.ResourceTemplates[i].Spec.Template.GroupVersionKind() != temp.Spec.Template.GroupVersionKind() {
-			continue
+// KubeAppApplyOption ensures resource templates are merged instead of replaced
+// before patch if they have the same name and GroupVersionKind.
+func KubeAppApplyOption() resource.ApplyOption {
+	return func(_ context.Context, current, desired runtime.Object) error {
+		c, ok := current.(*workloadv1alpha1.KubernetesApplication)
+		if !ok {
+			return errors.New(errNotKubeApp)
 		}
-		if desired.Spec.ResourceTemplates[i].GetName() != temp.GetName() {
-			continue
+		d, ok := desired.(*workloadv1alpha1.KubernetesApplication)
+		if !ok {
+			return errors.New(errNotKubeApp)
 		}
-		if err := mergo.Merge(desired.Spec.ResourceTemplates[i].Spec.Template, temp.Spec.Template); err != nil {
-			return errors.Wrap(err, errMergeKubeAppTemplates)
+		for i, temp := range c.Spec.ResourceTemplates {
+			if d.Spec.ResourceTemplates[i].Spec.Template.GroupVersionKind() != temp.Spec.Template.GroupVersionKind() {
+				continue
+			}
+			if d.Spec.ResourceTemplates[i].GetName() != temp.GetName() {
+				continue
+			}
```

Comment:
Are we relying on the current and desired objects both having an equal length, identically ordered slice of templates here? It seems like we might panic attempting to get `d.Spec.ResourceTemplates[i]` if the current kubeapp had  two templates, but the desired kubeapp had one (for example because we previously rendered a service, but no longer want to because its port was removed).

If this is an issue I think we could avoid it by mapping the name using a map of index, for example:

```go
type template struct {
    gvk  schema.GroupVersionKind
    name string
}

// Make a map of name and template GVK to index
index := make(map[template]int)
for i, t := range d.spec.ResourceTemplates {
    index[template{gvk: t.Spec.Template.GroupVersionKind(), name: t.GetName()}] = i
}

for _, t := range c.Spec.ResourceTemplates {
    i, ok := index[template{gvk: t.Spec.Template.GroupVersionKind(), name: t.GetName()}]

    // This template exists in the current KubeApp, but not the desired one. It will be removed because
    // resource.Apply uses a JSON merge patch, which overwrites the entire array.
    if !ok {
        continue
    }

    // This template exists in both the current KubeApp and the desired one. Merge the desired template
    // into the current one then make that our desired state.
    merge := []byte{}
    json.Unmarshal(merge, d.Spec.ResourceTemplates[i].Spec.Template)    
}
```

Reply:
Yep, you are spot on here, thanks for catching! I will also add a test case to demonstrate this.